### DISCUSSION TITLE
Labs Feedback: @lit-labs/observers

### DISCUSSION THREAD

**Discussion by AndrewJakubowicz** [lit/lit#3355] [Labs Feedback]
The `@lit-labs/observers` package provides a set of reactive controllers that facilitate using the platform observer objects, including:

 - MutationObserver
 - ResizeObserver
 - IntersectionObserver
 - PerformanceObserver

By managing these observers with a reactive controller, changes can be easily integrated into the Lit reactive update lifecycle. The controller can also help manage observer cleanup and rendering in response to changes.

We want to collect your feedback on your experiences with `@lit-labs/observers` as we iterate on it according to our Lit Labs process and move the package towards possible graduation or retirement.

- Status: **Under development**
- Issues: [[labs/observers]](https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+%5Blabs%2Fobservers%5D)
- Docs: [README](https://github.com/lit/lit/tree/main/packages/labs/observers#readme)
- npm: [@lit-labs/observers](https://www.npmjs.com/package/@lit-labs/observers)

Please try `@lit-labs/observers` and tell us about your experience. Leave comments with your thoughts; highlight what issues or missing features you're blocked on; give a link to a project that uses it; etc.

And so we can have some numbers to look at, you can add reactions to this comment to indicate your experience with this package:

🚀 I have tried this package and it works great. I need it. Ship it!
👍 I have tried this package and it seems like a good idea.
😕 I have tried this package and would use it, but it has serious bugs or missing features.
👀 I am interested, but haven't tried the package yet.
👎 I'm not interested in this package.

Note: other reactions like 😄, 🎉, and ❤️ don't mean anything as part of this scheme. You probably mean, and please use, 🚀, 👍, or  👀.

**Comment by cboehme** [comment]
We use `MutationController` and `IntersectionController` in one of our components. Both do a great job at integrating the observers without much boilerplate.

There was only one small feature that we missed: That an `unobserveAll()` method, which unobserves all currently watched elements. 

Why would this method help in our case? - Our component has a slot which takes a list of elements. We use `MutationController` and `IntersectionController` to watch for changes in these slotted elements. The `slotchange`-Event is used to update the list of observed elements. The currently slotted elements are retrieved via `@queryAssignedElements`. However, this list only contains the current elements not the ones that were removed from the slot. In order to unobserve these in the event handler, we need to keep our own list of observed elements in order to remove the ones being gone. As `MutationController` and `IntersectionController` manage such listes already, it would be handy to make use of these by providing an `unobserveAll` method.

I am happy to contribute these methods if there is interest in adding them.
